### ISSUE TITLE
Coroutines context does not change after operation is completed

### ISSUE THREAD

**Issue by satyamagarwal** [arrow-kt/arrow-fx#304]
**After operation is completed in `evalOn`, or `parTupledN` for example, the coroutine context should switch back to the parent, it does not happen. It also doesn't copy MDC contextMap from parent even when using MDCContext() coroutine context.**

This test illustrates the issue : 

https://github.com/satyamagarwal/arrow-issues/blob/master/src/test/kotlin/com/arrow/issues/CoroutinesTest.kt

It has been tested with `Environment` from `arrow-fx-coroutines` and produces same result.

Discussion : 
https://kotlinlang.slack.com/archives/C5UPMM0A0/p1605473258437200

**Comment by nomisRev** [comment]
> After operation is completed in evalOn, or parTupledN for example, the coroutine context should switch back to the parent, it does not happen.

This can be considered a bug since `Environment` actually requires a dispatching `ContinuationInterceptor` as an argument, which `MDCContext` isn't. So we can default back to adding `ComputationPool` by default if the provided `CoroutineContext` isn't a `ContinuationInterceptor`. That would fix the incorrect thread reading.

The reason for this is that to go back to an original thread you need a dispatching `ContinuationInterceptor`, otherwise, that's not possible.

> It also doesn't copy MDC contextMap from parent even when using `MDCContext` coroutine context.

This is a bit of a different story, since `MDCContext` is a class specifically built on top of Kotlin Coroutines's [`ThreadContextLocal`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-thread-context-element/), which is hard-wired its operations to maintain the `contextMap` which normally happens with regular `ThreadLocal` if I understand `MDC` correctly.
AFAIK It's impossible for Arrow Fx Coroutines to interop-opt with this behavior. 

@satyamagarwal could you try changing the following lines, and running the tests again?

- [`Environment(MDCContext()).unsafeRunSync` => `Enviroment(Dispatchers.Default + MDCContext())`](https://github.com/satyamagarwal/arrow-issues/blob/27cb0dc1cac127090b9863439d6532774f5f1619/src/test/kotlin/com/arrow/issues/CoroutinesTest.kt#L94)

- Change all occurences of `ComputationPool + MDCContext()` to `Dispatchers.Default + MDCContext()`

I'm a bit confused about the last two test cases in that file, I need to make some time to run them locally and check what's going on there.

**Comment by nomisRev** [comment]
@satyamagarwal I did the test locally, and it seems relying on KotlinX Coroutines dispatchers works. Not 100% sure if it covers all the use-cases of correctly tho.

**Comment by satyamagarwal** [comment]
@nomisRev Thank you. 

I understand about the MDCContext. lets look at this test : 

https://github.com/satyamagarwal/arrow-issues/blob/master/src/test/kotlin/com/arrow/issues/CoroutinesTest.kt#L156-L184

Lets change https://github.com/satyamagarwal/arrow-issues/blob/master/src/test/kotlin/com/arrow/issues/CoroutinesTest.kt#L167 to 

```
val threadName: String = evalOn(Dispatchers.Default) {
```

Now, this produces : 

```
I am in main @coroutine#3
I am in parTupledN, DefaultDispatcher-worker-1
I am in parTupledN, DefaultDispatcher-worker-2
I am in main @coroutine#3
I am in DefaultDispatcher-worker-2
Evaluated thread result : DefaultDispatcher-worker-2 in DefaultDispatcher-worker-2
I am in ForkJoinPool-1-worker-5
I am in ForkJoinPool-1-worker-3
Evaluated either block result : DefaultDispatcher-worker-2 in ForkJoinPool-1-worker-3
```

But I am expecting it to print this : 

```
I am in main @coroutine#3
I am in parTupledN, DefaultDispatcher-worker-1
I am in parTupledN, DefaultDispatcher-worker-2
I am in main @coroutine#3
I am in DefaultDispatcher-worker-2
Evaluated thread result : DefaultDispatcher-worker-2 in **@coroutine#3**
I am in ForkJoinPool-1-worker-5
I am in ForkJoinPool-1-worker-3
Evaluated either block result : DefaultDispatcher-worker-2 in **@coroutine#3**
```

Is my expectation correct ?

**Comment by nomisRev** [comment]
Your expectation is correct, and this is also unrelated to `MDCContext`. I'll have to investigate, this is only happening with `kotlinx.coroutines.runBlocking` and not with `arrow.fx.coroutines.Enviroment()`?

**Comment by satyamagarwal** [comment]
Yes pretty much. This program : 

```
@Test
fun `it just prints the different threads the cursor is in without MDCContext with Arrow's Environment`(): Unit = Environment(Dispatchers.Unconfined).unsafeRunSync {
    println("I am in ${getThreadName()}")

    parTupledN(
        Dispatchers.Default,
        suspend { println("I am in parTupledN, ${getThreadName()}") },
        suspend { println("I am in parTupledN, ${getThreadName()}") }
    )

    val result: Either<Throwable, String> = either {
        println("I am in ${getThreadName()}")

        val threadName: String = evalOn(Dispatchers.Default) {
            println("I am in ${getThreadName()}")

            getThreadName().right()
        }.bind()

        println("Evaluated thread result : $threadName in ${getThreadName()}")

        parTupledN(
            suspend { println("I am in ${getThreadName()}") },
            suspend { println("I am in ${getThreadName()}") }
        )

        threadName
    }

    println("Evaluated either block result : ${result.orNull()} in ${getThreadName()}")
}
```

Prints : 
```
I am in main
I am in parTupledN, DefaultDispatcher-worker-1
I am in parTupledN, DefaultDispatcher-worker-2
I am in DefaultDispatcher-worker-2
I am in DefaultDispatcher-worker-2
Evaluated thread result : DefaultDispatcher-worker-2 in DefaultDispatcher-worker-2
I am in DefaultDispatcher-worker-1
I am in DefaultDispatcher-worker-1
Evaluated either block result : DefaultDispatcher-worker-2 in DefaultDispatcher-worker-1
```

But I am expecting : 
```
I am in main
I am in parTupledN, DefaultDispatcher-worker-1
I am in parTupledN, DefaultDispatcher-worker-2
I am in **main**
I am in DefaultDispatcher-worker-2
Evaluated thread result : DefaultDispatcher-worker-2 in **main**
I am in DefaultDispatcher-worker-1
I am in DefaultDispatcher-worker-1
Evaluated either block result : DefaultDispatcher-worker-2 in **main**
```

What do you think ? @nomisRev

**Comment by nomisRev** [comment]
That is expected... Since `Dispatchers.Unconfined` is not a dispatching `ContinuationInterceptor`. Afaik, it's a kind of special marker for KotlinX Coroutines specific behavior, but it can't actually do any dispatching.

So what you're seeing there is that once it goes somewhere (main -> Dispatchers.Defualt), it cannot come back to **main** since how to go back to **main** is "undefined". Or rather it's unknown how to go back to **main**. If you'd pass `IOPool` to `Enviroment()` there, you'll see it work with the expected behavior.

What `kotlinx.coroutines.runBlocking` does is launch an `EventLoop` in the case there is no `ContinuationInterceptor`, and they use that to go back to the original thread but that's complex and expensive machinery we don't have.

As mentioned before:
> This can be considered a bug since Environment actually requires a dispatching ContinuationInterceptor as an argument, which MDCContext isn't. So we can default back to adding ComputationPool by default if the provided CoroutineContext isn't a ContinuationInterceptor. That would fix the incorrect thread reading.
>
> The reason for this is that to go back to an original thread you need a dispatching ContinuationInterceptor, otherwise, that's not possible.

**Comment by satyamagarwal** [comment]
Ok, I think I understand, now. But then it means the bug lies in `parTupledN` and `evalOn` implementations correct ? 

So I ran this test : 

```
@Test
fun `it just prints the different threads the cursor is in without MDCContext with Arrow's Environment`() {
    Environment(IOPool).unsafeRunSync {
        println("I am in ${getThreadName()}")

        parTupledN(
            Dispatchers.Default,
            suspend { println("I am in parTupledN, ${getThreadName()}") },
            suspend { println("I am in parTupledN, ${getThreadName()}") }
        )

        val result: Either<Throwable, String> = either {
            println("I am in ${getThreadName()}")

            val threadName: String = evalOn(Dispatchers.Default) {
                println("I am in ${getThreadName()}")

                getThreadName().right()
            }.bind()

            println("Evaluated thread result : $threadName in ${getThreadName()}")

            parTupledN(
                ComputationPool,
                suspend { println("I am in ${getThreadName()}") },
                suspend { println("I am in ${getThreadName()}") }
            )

            threadName
        }

        println("Evaluated either block result : ${result.orNull()} in ${getThreadName()}")
    }
}
```
It prints : 
```
I am in io-arrow-kt-worker-0
I am in parTupledN, DefaultDispatcher-worker-1
I am in parTupledN, DefaultDispatcher-worker-2
I am in io-arrow-kt-worker-0
I am in DefaultDispatcher-worker-1
Evaluated thread result : DefaultDispatcher-worker-1 in DefaultDispatcher-worker-1
I am in ForkJoinPool-1-worker-3
I am in ForkJoinPool-1-worker-5
Evaluated either block result : DefaultDispatcher-worker-1 in ForkJoinPool-1-worker-5
```
But I am expecting : 
```
I am in io-arrow-kt-worker-0
I am in parTupledN, DefaultDispatcher-worker-1
I am in parTupledN, DefaultDispatcher-worker-2
I am in io-arrow-kt-worker-0
I am in DefaultDispatcher-worker-1
Evaluated thread result : DefaultDispatcher-worker-1 in **io-arrow-kt-worker-0**
I am in ForkJoinPool-1-worker-3
I am in ForkJoinPool-1-worker-5
Evaluated either block result : DefaultDispatcher-worker-1 in **io-arrow-kt-worker-0**
```

This expectation is also correct ? @nomisRev

**Comment by nomisRev** [comment]
@satyamagarwal your expectation is correct! Thanks for helping narrow down the issue, going to write a test, and look into it tomorrow. I'll keep you posted here if I find anything 👍

**Comment by nomisRev** [comment]
@satyamagarwal all these issues should be fixed with #317, since we now rely on KotlinX you can also safely use all integrations for KotlinX without any weird bugs such as `MDContext`! Thanks for reporting and support 👏 👏
